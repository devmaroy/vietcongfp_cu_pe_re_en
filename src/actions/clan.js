// Clan Actions

import database from '../firebase/firebase';
import history from '../routers/history';
import { 
    GET_CLANS,
    GET_CLAN,
    ADD_CLAN,
    EDIT_CLAN,
    REMOVE_CLAN,
    SET_CLAN_LOADING,
    CLEAR_CLAN_LOADING,
    GET_ERRORS
} from './types';


// Get Clans
export const startGetClans = () => {
    return ( dispatch ) => {
        dispatch( setClanLoading() );

        return database
            .ref( 'clans' )
            .orderByChild( 'createdAt' )
            .once( 'value' )
            .then( ( snapshot ) => {
                const clans = [];

                snapshot.forEach( ( childSnapshot ) => {
                    clans.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });

                dispatch({
                    type: GET_CLANS,
                    clans
                });
    
                dispatch( clearClanLoading() );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Get Clan
export const startGetClanById = ( id ) => {
    return ( dispatch ) => {
        dispatch( setClanLoading() );

        return database
            .ref( `clans/${ id }` )
            .once( 'value' )
            .then( ( snapshot ) => {
                if ( snapshot.exists() ) {
                    const clan = {
                        id: snapshot.key,
                        ...snapshot.val()
                    };
        
                    dispatch({
                        type: GET_CLAN,
                        clan
                    });
    
                    dispatch( clearClanLoading() );
                } else {
                    // Simple redirect to non existent route
                    history.push( '/not-found' );
                }
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Add Clan
export const startAddClan = ( clanData ) => {
    return ( dispatch ) => {
        return database
            .ref( 'clans' )
            .push( clanData )
            .then( ( ref ) => {
                dispatch({
                    type: ADD_CLAN,
                    clan: {
                        id: ref.key,
                        ...clanData 
                    }
                });
    
                history.push( '/clans' );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
} ;


// Edit Clan 
export const startEditClan = ( id, updates ) => {
    return ( dispatch ) => {
        return database
            .ref( `clans/${ id }` )
            .update( updates )
            .then( ( ref ) => {
                dispatch({
                    type: EDIT_CLAN,
                    id,
                    updates
                });
    
                history.push( '/clans' );
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Remove Clan
export const startRemoveClan = ( id ) => {
    return ( dispatch ) => {
        return database 
            .ref( `clans/${ id }` )
            .remove()
            .then( () => {
                dispatch({
                    type: REMOVE_CLAN,
                    id
                });
            })
            .catch( ( err ) => {
                // Simple redirect to non existent route
                history.push( '/oops' );
            });
    };
};


// Clan Loading
export const setClanLoading = () => ({
    type: SET_CLAN_LOADING
});


// Clear Clan Loading
export const clearClanLoading = () => ({
    type: CLEAR_CLAN_LOADING
});