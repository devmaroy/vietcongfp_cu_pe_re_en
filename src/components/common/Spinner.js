import React from 'react';
import PropTypes from 'prop-types';

const Spinner = ( { imagePath, alt, style, classNameWrapper, classNameImg } ) => (
    <div className={ classNameWrapper }>
        <img 
            style={ style }
            src={ imagePath } 
            alt={ alt }
            className={ classNameImg }
        />
    </div>
);

Spinner.propTypes = {
    imagePath: PropTypes.string,
    alt: PropTypes.string,
    style: PropTypes.object,
    classNameWrapper: PropTypes.string,
    classNameImg: PropTypes.string
};

Spinner.defaultProps = {
    imagePath: '/images/loader.gif',
    alt: 'Loading...',
    style: { width: '80px', margin: '60px', display: 'block' }
};

export default Spinner;