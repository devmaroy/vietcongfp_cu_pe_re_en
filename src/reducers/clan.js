// Clan Reducer

import {
    GET_CLANS,
    GET_CLAN,
    ADD_CLAN,
    EDIT_CLAN,
    REMOVE_CLAN,
    SET_CLAN_LOADING,
    CLEAR_CLAN_LOADING
} from '../actions/types';

const clanReducerDefaultState = {
    clans: [],
    clan: {},
    isClanLoading: false
};

export default ( state = clanReducerDefaultState, action ) => {
    switch ( action.type ) {
        
        case GET_CLANS:
            return {
                ...state,
                clans: action.clans.reverse()
            };

        case GET_CLAN:
            return {
                ...state,
                clan: action.clan
            };

        case ADD_CLAN:
            return {
                ...state,
                clans: [ action.clan, ...state.clans ]
            };

        case EDIT_CLAN:
            return {
                ...state,
                clans: state.clans.map( ( clan ) => {
                    if ( clan.id === action.id ) {
                        return {
                            ...clan,
                            ...action.updates
                        }
                    } else {
                        return clan;
                    }
                })
            };

        case REMOVE_CLAN:
            return {
                ...state,
                clans: state.clans.filter( ( clan ) => clan.id !== action.id )
            };
        
        case SET_CLAN_LOADING:
            return {
                ...state,
                isClanLoading: true
            };

        case CLEAR_CLAN_LOADING:
            return {
                ...state,
                isClanLoading: false
            };
    
        default:
            return state;
    };
};