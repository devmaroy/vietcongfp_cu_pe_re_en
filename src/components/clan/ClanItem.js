import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startRemoveClan } from '../../actions/clan';
import { Link } from 'react-router-dom';
import { FlagIcon } from "react-flag-kit";

export const ClanItem = ( props ) => {
    const { clan, isAuthenticated, startRemoveClan } = props;

    const onRemove = ( id ) => {
        startRemoveClan( id );
    };

    return (
        <tr className="clan">
            <td>
                <FlagIcon 
                    code={ clan.country } 
                    size={ 26 } 
                    title={ `Country flag ${ clan.country }` }
                />
            </td>
            <td>{ clan.tag }</td>
            <td>{ clan.name }</td>
            <td>{ clan.leader }</td>
            <td>
                <a href={ clan.url }>{ clan.url }</a>
            </td>
            <td>
                { ( clan.eslName && clan.eslUrl ) ? <a href={ clan.eslUrl }>{ clan.eslName }</a> : 'no' }
            </td>
            { 
                isAuthenticated && (
                    <td>
                        <Link to={ `/clans/${ clan.id }/edit` }>
                            <img src="/images/icons/edit.svg" alt="Edit icon" className="icon edit-icon" />
                        </Link>
                        <button onClick={ onRemove.bind( this, clan.id ) }>
                            <img src="/images/icons/remove.svg" alt="Remove icon" className="icon remove-icon"/>
                        </button>
                    </td>
                )
            }
        </tr>
    )
};

ClanItem.propTypes = {
    clan: PropTypes.object.isRequired,
    isAuthenticated: PropTypes.bool,
    startRemoveClan: PropTypes.func.isRequired
}

const mapStateToProps = ( state ) => ({
    isAuthenticated: state.auth.isAuthenticated
});

const mapDispatchToProps = ( dispatch ) => ({
    startRemoveClan: ( id ) => dispatch( startRemoveClan( id ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( ClanItem );