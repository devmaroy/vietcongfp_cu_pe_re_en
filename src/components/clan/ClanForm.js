import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import TextFieldGroup from '../common/TextFieldGroup';
import TextAreaFieldGroup from '../common/TextAreaFieldGroup';
import SelectFieldGroup from '../common/SelectFieldGroup';
import validator from 'validator';
import Spinner from '../common/Spinner';
import moment from 'moment';
import countries from '../../utils/countries';

export class ClanForm extends React.Component {
    state = {
        tag: this.props.clan ? this.props.clan.tag : '',
        name: this.props.clan ? this.props.clan.name : '',
        country: this.props.clan ? this.props.clan.country : '',
        leader: this.props.clan ? this.props.clan.leader : '',
        url: this.props.clan ? this.props.clan.url : '',
        eslName: this.props.clan ? this.props.clan.eslName : '',
        eslUrl: this.props.clan ? this.props.clan.eslUrl : '',
        createdAt: this.props.clan ? this.props.clan.createdAt : moment(),
        errors: {}
    };

    handleValidation = () => {
        const errors = {};
        let formIsValid = true;

        // Tag
        if ( ! this.state.tag ) {
            formIsValid = false;
            errors[ 'tag' ] = 'Tag cannot be empty';
        }

        // Name
        if ( ! this.state.name ) {
            formIsValid = false;
            errors[ 'name' ] = 'Name cannot be empty';
        }

        // Country
        if ( ! this.state.country ) {
            formIsValid = false;
            errors[ 'country' ] = 'Country must be provided';
        }

        // Leader
        if ( ! this.state.leader ) {
            formIsValid = false;
            errors[ 'leader' ] = 'Leader cannot be empty';
        }

        // Url
        if ( ! this.state.url ) {
            formIsValid = false;
            errors[ 'url' ] = 'Url cannot be empty';
        } else if ( ! validator.isURL( this.state.url )  ) {
            formIsValid = false;
            errors[ 'url' ] = 'Url must be valid';
        }

        // ESL Url
        if ( this.state.eslUrl ) {
            if ( ! validator.isURL( this.state.eslUrl ) ) {
                formIsValid = false;
                errors[ 'eslUrl' ] = 'ESL url must be valid';
            }
        }

        this.setState( () => ( { errors } ) );

        return formIsValid;
    };

    handleChange = ( e ) => {
        const target = e.target;
        const value = target.value;
        const name = target.name;

        this.setState( () => ( { [ name ]: value } ) );
    };

    onSubmit = ( e ) => {
        e.preventDefault();

        if ( this.handleValidation() ) {
            const clanData = {
                tag: this.state.tag,
                name: this.state.name,
                country: this.state.country,
                leader: this.state.leader,
                url: this.state.url,
                eslName: this.state.eslName,
                eslUrl: this.state.eslUrl,
                createdAt: this.state.createdAt.valueOf()
            };
            
            this.props.onSubmit( clanData );
        }
    }

    render() {
        const { errors } = this.state;

        return (
            <form onSubmit={ this.onSubmit } className="form" noValidate>
                <div className="input-group">
                    <div className="input-group__item">
                        <TextFieldGroup 
                            name="tag"
                            placeholder="Tag"
                            value={ this.state.tag }
                            onChange={ this.handleChange }
                            error={ errors.tag }
                            className="text-input"
                        />
                    </div>
                </div>
                <div className="input-group">
                    <div className="input-group__item">
                        <TextFieldGroup 
                            name="name"
                            placeholder="Name"
                            value={ this.state.name }
                            onChange={ this.handleChange }
                            error={ errors.name }
                            className="text-input"
                        />
                    </div>
                </div>
                <div className="input-group">
                    <div className="input-group__item">
                        <SelectFieldGroup
                            name="country"
                            value={ this.state.country }
                            options={ countries }
                            onChange={ this.handleChange }
                            className="select"
                        />
                    </div>
                </div>
                <div className="input-group">
                    <div className="input-group__item">
                        <TextFieldGroup 
                            name="leader"
                            placeholder="Leader"
                            value={ this.state.leader }
                            onChange={ this.handleChange }
                            error={ errors.leader }
                            className="text-input"
                        />
                    </div>
                </div>
                <div className="input-group">
                    <div className="input-group__item">
                        <TextFieldGroup 
                            name="url"
                            type="url"
                            placeholder="Website"
                            value={ this.state.url }
                            onChange={ this.handleChange }
                            error={ errors.url }
                            className="text-input"
                        />
                    </div>
                </div>
                <div className="input-group">
                    <div className="input-group__item">
                        <TextFieldGroup 
                            name="eslName"
                            placeholder="ESL Clan name"
                            value={ this.state.eslName }
                            onChange={ this.handleChange }
                            error={ errors.eslName }
                            className="text-input"
                        />
                    </div>
                </div>
                <div className="input-group">
                    <div className="input-group__item">
                        <TextFieldGroup 
                            name="eslUrl"
                            type="url"
                            placeholder="ESL Clan website"
                            value={ this.state.eslUrl }
                            onChange={ this.handleChange }
                            error={ errors.eslUrl }
                            className="text-input"
                        />
                    </div>
                </div>
                <div className="form__meta">
                    <button className="button">Save Clan</button>
                </div>
            </form>
        );
    };
};

ClanForm.propTypes = {
    clan: PropTypes.object,
    onSubmit: PropTypes.func.isRequired
};

export default ClanForm;