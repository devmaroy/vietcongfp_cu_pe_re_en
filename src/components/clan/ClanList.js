import React from 'react';
import PropTypes from 'prop-types';
import Pagination from "react-js-pagination";
import ClanItem from './ClanItem';

class ClanList extends React.Component {
    state = {
        activePage: 1,
        clansPerPage: 5
    }

    handlePageChange = ( pageNumber ) => {
        this.setState( () => ( { activePage: pageNumber } ) );
    }

    render() {
        const { activePage, clansPerPage } = this.state;
        const { clans } = this.props;

        return (
            <div>
                {
                    clans.length === 0 ? (
                        <p>There are no clans.</p>
                    ) : (
                        <div>
                            <table className="clans">
                                <thead>
                                    <tr>
                                        <th>Country</th>
                                        <th>Tag</th>
                                        <th>Name</th>
                                        <th>Leader</th>
                                        <th>Web</th>
                                        <th>ESL</th>
                                    </tr>
                                </thead>
                                <tbody>
                                    {
                                        clans
                                            .slice( ( activePage - 1 ) * clansPerPage, ( clansPerPage * activePage ) )
                                            .map( ( clan ) => <ClanItem key={ clan.id } clan={ clan } /> )
                                    }
                                </tbody>
                            </table>
                            {
                                clans.length > clansPerPage && (
                                    <div className="site-pagination">
                                        <Pagination
                                            activePage={ this.state.activePage }
                                            itemsCountPerPage={ clansPerPage }
                                            totalItemsCount={ clans.length }
                                            pageRangeDisplayed={ 5 }
                                            onChange={ this.handlePageChange }
                                        />
                                    </div>
                                )
                            }
                        </div>
                    )
                }
            </div>
        );
    };
};

ClanList.propTypes = {
    clans: PropTypes.array.isRequired
};

export default ClanList;