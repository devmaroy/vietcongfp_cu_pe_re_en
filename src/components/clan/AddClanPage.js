import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startAddClan } from '../../actions/clan';
import ClanForm from './ClanForm';

export class AddClanPage extends React.Component {
    onSubmit = ( clanData ) => {
        this.props.startAddClan( clanData ); 
    };

    render() {
        return (
            <div className="content-container">
                <div className="page">
                    <h2>Create Clan</h2>
                    <ClanForm 
                        onSubmit={ this.onSubmit }
                    />
                </div>
            </div>
        )
    };
};

AddClanPage.propTypes = {
    startAddClan: PropTypes.func.isRequired
};

const mapDispatchToProps = ( dispatch ) => ({
    startAddClan: ( clanData ) => dispatch( startAddClan( clanData ) )
});

export default connect( undefined, mapDispatchToProps )( AddClanPage );