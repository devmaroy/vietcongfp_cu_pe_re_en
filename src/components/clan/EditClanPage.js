import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { startGetClanById, startEditClan } from '../../actions/clan';
import Spinner from '../common/Spinner';
import ClanForm from './ClanForm';

export class EditClanPage extends React.Component {
    componentDidMount() {
        this.props.startGetClanById( this.props.match.params.id );
    };

    onSubmit = ( updates ) => {
        this.props.startEditClan( this.props.match.params.id, updates );
    };

    render() {
        const { clan, isClanLoading } = this.props.clan;
        let clanEditContent;

        if ( clan === null || isClanLoading || Object.keys( clan ).length === 0 ) {
            clanEditContent = <Spinner />;
        } else {
            clanEditContent = (
               <div>
                    <h2>Edit Clan</h2>
                    <ClanForm
                        onSubmit={ this.onSubmit }
                        clan={ clan }
                    />
               </div>
           );
        }

        return (
            <div className="content-container">
                <div className="page">
                    { clanEditContent }
                </div>
            </div>
        )
    }
}

EditClanPage.propTypes = {
    startGetClanById: PropTypes.func.isRequired,
    startEditClan: PropTypes.func.isRequired,
    clan: PropTypes.object.isRequired
};

const mapStateToProps = ( state ) => ({
    clan: state.clan
});

const mapDispatchToProps = ( dispatch ) => ({
    startGetClanById: ( id ) => dispatch( startGetClanById( id ) ),
    startEditClan: ( id, updates ) => dispatch( startEditClan( id, updates ) )
});

export default connect( mapStateToProps, mapDispatchToProps )( EditClanPage );