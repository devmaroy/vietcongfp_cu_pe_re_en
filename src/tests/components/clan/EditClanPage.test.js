import React from 'react';
import { shallow } from 'enzyme';
import { EditClanPage } from '../../../components/clan/EditClanPage';
import ClanForm from '../../../components/clan/ClanForm';
import Spinner from '../../../components/common/Spinner';
import clans from '../../fixtures/clans';

test( 'should render correctly EditClanPage with spinner', () => {
    const clan = {
        clan: {},
        isClanLoading: true
    }

    const wrapper = shallow(
        <EditClanPage
            startGetClanById={ () => {} }
            startEditClan={ () => {} }
            clan={ clan }
            match={ { params: { id: '1' } } }
        />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( Spinner ).exists() ).toBeTruthy();
});

test( 'should render correctly EditClanPage without spinner', () => {
    const clan = {
        clan: clans[0],
        isClanLoading: false
    }

    const wrapper = shallow(
        <EditClanPage
            startGetClanById={ () => {} }
            startEditClan={ () => {} }
            clan={ clan }
            match={ { params: { id: '1' } } }
        />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( Spinner ).exists() ).toBeFalsy();
});

test( 'should render ClanForm correctly', () => {
    const clan = {
        clan: clans[0],
        isClanLoading: false
    }

    const wrapper = shallow(
        <EditClanPage
            startGetClanById={ () => {} }
            startEditClan={ () => {} }
            clan={ clan }
            match={ { params: { id: '1' } } }
        />
    );

    expect( wrapper ).toMatchSnapshot();
    expect( wrapper.find( ClanForm ).exists() ).toBeTruthy();
});

test( 'should call startGetClan', () => {
    const clan = {
        clan: clans[0],
        isClanLoading: false
    }

    const startGetClanById = jest.fn();

    const wrapper = shallow(
        <EditClanPage
            startGetClanById={ startGetClanById }
            startEditClan={ () => {} }
            clan={ clan }
            match={ { params: { id: '1' } } }
        />
    );

    expect( startGetClanById ).toHaveBeenCalled();
});

test( 'should handle startEditClan', () => {
    const startEditClan = jest.fn();

    const clan = {
        clan: clans[1],
        isClanLoading: false
    };

    const wrapper = shallow(
        <EditClanPage
            startGetClanById={ () => {} }
            startEditClan={ startEditClan }
            clan={ clan }
            match={ { params: { id: clans[1].id } } }
        />
    );

    wrapper.find( ClanForm ).prop( 'onSubmit' )( clans[1] );

    expect( startEditClan ).toHaveBeenLastCalledWith( clans[1].id, clans[1] );
});