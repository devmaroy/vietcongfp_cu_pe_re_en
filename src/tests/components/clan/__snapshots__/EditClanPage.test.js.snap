// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ClanForm correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <div>
      <h2>
        Edit Clan
      </h2>
      <ClanForm
        clan={
          Object {
            "country": "CZ",
            "createdAt": 1000,
            "eslName": "Dog Soldiers",
            "eslUrl": "https://esl.com/team/110",
            "id": "1",
            "leader": "Lue09",
            "name": "Dog Soldiers",
            "tag": "doG",
            "url": "https://dogsoldiers.cz/",
          }
        }
        onSubmit={[Function]}
      />
    </div>
  </div>
</div>
`;

exports[`should render correctly EditClanPage with spinner 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <Spinner
      alt="Loading..."
      imagePath="/images/loader.gif"
      style={
        Object {
          "display": "block",
          "margin": "60px",
          "width": "80px",
        }
      }
    />
  </div>
</div>
`;

exports[`should render correctly EditClanPage without spinner 1`] = `
<div
  className="content-container"
>
  <div
    className="page"
  >
    <div>
      <h2>
        Edit Clan
      </h2>
      <ClanForm
        clan={
          Object {
            "country": "CZ",
            "createdAt": 1000,
            "eslName": "Dog Soldiers",
            "eslUrl": "https://esl.com/team/110",
            "id": "1",
            "leader": "Lue09",
            "name": "Dog Soldiers",
            "tag": "doG",
            "url": "https://dogsoldiers.cz/",
          }
        }
        onSubmit={[Function]}
      />
    </div>
  </div>
</div>
`;
